endpoints/openrtb2/auction.go

```diff
@@ -56,24 +62,56 @@ func (deps *endpointDeps) Auction(w http.ResponseWriter, r *http.Request, _ http
 	w.Write(responseBytes)
 }
 
-// parseRequest turns the HTTP request into an OpenRTB request.
+// parseRequest turns the HTTP request into an OpenRTB request. This is guaranteed to return:
 //
-// This will return an error if the request couldn't be parsed, or if the request isn't valid according
-// to the OpenRTB 2.5 spec.
+//   - A context which times out appropriately, given the request.
+//   - A cancellation function which should be called if the auction finishes early.
 //
-// It will also return errors for some of the "strong recommendations" in the spec, as long as
-// the same request can be sent in a better way which agrees with the recommendations.
-func (deps *endpointDeps) parseRequest(httpRequest *http.Request) (*openrtb.BidRequest, error) {
-	var ortbRequest openrtb.BidRequest
-	if err := json.NewDecoder(httpRequest.Body).Decode(&ortbRequest); err != nil {
-		return nil, err
+// If the errors list is empty, then the returned request will be valid according to the OpenRTB 2.5 spec.
+// In case of "strong recommendations" in the spec, it tends to be restrictive. If a better workaround is
+// possible, it will return errors with messages that suggest improvements.
+//
+// If the errors list has at least one element, then no guarantees are made about the returned request.
+func (deps *endpointDeps) parseRequest(httpRequest *http.Request) (req *openrtb.BidRequest, ctx context.Context, cancel func(), errs []error) {
+	req = &openrtb.BidRequest{}
+	ctx = context.Background()
+	cancel = func() { }
+	errs = nil
+
+	// Pull the request body into a buffer, so we have it for later usage.
+	lr := &io.LimitedReader{ httpRequest.Body, deps.cfg.MaxRequestSize }
+	rawRequest, err := ioutil.ReadAll(lr)
+	if err != nil {
+		errs = []error{err}
+		return
+	}
+	// If the request size was too large, read through the rest of the request body so that the connection can be reused.
+	if lr.N <= 0 {
```

Comment:
I like how it also validates if `lr` was initialized with a bad size.

Note: `lr.N` would be 0 even when request size=`MaxRequestSize` and returns an error when it shouldn't. Don't know how much that could matter.

Reply:
Good catch! It's not likely to affect too many requests... but it's very serious when it occurs, because we'd return a 400 rather than running the auction.

I'll fix this.

Reply:
Great! thanks